File: 2015/Sep/prayatna.cpp
#include <iostream>
#include <cstdio>
#include <vector>
using namespace std;

const int MAX = 100001;


void dfs(vector<int> v[], bool visited[], int root)
{
    visited[root] = true;
    for (int i = 0; i < v[root].size(); ++i)
    {
        int adj = v[root][i];
        if (!visited[adj])
        {
            dfs(v, visited, adj);
        }
    }
}

int no_of_connected_components(vector<int> v[], int n)
{
    bool visited[n] = {false};
    int cnt = 0;
    for (int i = 0; i < n; ++i)
    {
        if (!visited[i])
        {
            dfs(v, visited, i);
            cnt++;
        }
    }
    return cnt;
}

int main()
{
    int tc;
    scanf("%d", &tc);
    while (tc--)
    {
        int n, m;
        scanf("%d%d", &n, &m);
        vector<int> v[n];
        while (m--)
        {
            int x, y;
            scanf("%d%d", &x, &y);
            v[x].push_back(y);
            v[y].push_back(x);
        }
        printf("%d\n", no_of_connected_components(v, n));
    }
    
    return 0;
}

File: 2015/Sep/prayatna.py
'''
Created on 24-Sep-2015

@author: Venkatesh
'''

from collections import defaultdict


def read_int_list():
    return [int(x) for x in raw_input().split()]


def read_int():
    return int(raw_input())


def no_of_connected_components(d, n):
    visited = [False] * n
    cnt = 0
    for v in xrange(n):
        if not visited[v]:
            dfs(d, visited, v)
            cnt += 1
    return cnt


def dfs(graph, visited, root):
    visited[root] = True
    for v in graph[root]:
        if not visited[v]:
            dfs(graph, visited, v)


def main():
    tc = read_int()
    for _ in xrange(tc):
        raw_input()  # Two blank lines
        raw_input()
        n = read_int()
        d = defaultdict(list)
        for _ in xrange(read_int()):
            x, y = read_int_list()
            d[x].append(y)
            d[y].append(x)
        print no_of_connected_components(d, n)


if __name__ == '__main__':
    main()
